=== modified file 'parser/rc.apparmor.functions'
--- parser/rc.apparmor.functions~	2020-12-02 12:01:37.000000000 +0100
+++ parser/rc.apparmor.functions	2021-02-07 15:19:11.388572849 +0100
@@ -32,7 +32,7 @@
 # Some nice defines that we use
 
 PARSER=/sbin/apparmor_parser
-PARSER_OPTS=
+PARSER_OPTS="-W"
 # Suppress warnings when booting in quiet mode
 if [ "${QUIET:-no}" = yes ] || [ "${quiet:-n}" = y ]; then
 	PARSER_OPTS="$PARSER_OPTS --quiet"

