Index: parser/parser_include.c
===================================================================
--- parser/parser_include.c	(wersja 246)
+++ parser/parser_include.c	(wersja 247)
@@ -393,8 +393,13 @@
 	}
 
 	if (S_ISDIR(my_stat.st_mode)) {
-		DIR *dir = fdopendir(fileno(newf));
+		DIR *dir = NULL;
 		struct dirent *dirent;
+
+		/* XXX - fdopendir not available in glibc < 2.4 */
+		/* dir = fdopendir(fileno(newf)); */
+		fclose(newf);
+		dir = opendir(buf);
 		if (!dir) {
 			retval = 1;
 			goto out;
