--- parser.org/Makefile	2010-03-11 08:28:38.000000000 +0100
+++ parser/Makefile	2010-03-16 22:42:34.830028680 +0100
@@ -125,9 +125,12 @@
 techdoc.txt: techdoc/index.html
 	w3m -dump $< > $@
 
-all: 	$(TOOLS) $(MANPAGES) ${HTMLMANPAGES} techdoc.pdf
+main: 	$(TOOLS)
 	$(Q)make -C po all
-	$(Q)make -s tests
+
+docs:	$(MANPAGES) ${HTMLMANPAGES}
+
+all:	main docs tests
 
 apparmor_parser: $(OBJECTS) $(PCREOBJECTS) $(AAREOBJECTS)
 	rm -f ./libstdc++.a
